cookbooks/prescribed_velocity_ascii_data/prescribed_velocity_ascii_data.cc

```diff
@@ -0,0 +1,350 @@
+                      << "\t'" << prm.get("Function expression") << "'";
+            throw;
+          }
+      }
+      prm.leave_subsection();
+    }
+    prm.leave_subsection ();
+
+    prescribed_velocity_ascii_file_loaded = false;
+  }
+
+
+
+  /**
+   * This function retrieves the unit support points (in the unit cell) for the current element.
+   * The DGP element used when 'set Use locally conservative discretization = true' does not
+   * have support points. If these elements are in use, a fictitious support point at the cell
+   * center is returned for each shape function that corresponds to the pressure variable,
+   * whereas the support points for the velocity are correct; the fictitious points don't matter
+   * because we only use this function when interpolating the velocity variable, and ignore the
+   * evaluation at the pressure support points.
+   */
+  template <int dim>
+  std::vector< Point<dim> >
+  get_unit_support_points_for_velocity(const SimulatorAccess<dim> &simulator_access)
+  {
+    std::vector< Point<dim> > unit_support_points;
+    if ( !simulator_access.get_parameters().use_locally_conservative_discretization )
+      {
+        return simulator_access.get_fe().get_unit_support_points();
+      }
+    else
+      {
+        //special case for discontinuous pressure elements, which lack unit support points
+        std::vector< Point<dim> > unit_support_points;
+        const unsigned int dofs_per_cell = simulator_access.get_fe().dofs_per_cell;
+        for (unsigned int dof=0; dof < dofs_per_cell; ++dof)
+          {
+            // base will hold element, base_index holds node/shape function within that element
+            const unsigned int
+            base       = simulator_access.get_fe().system_to_base_index(dof).first.first,
+            base_index = simulator_access.get_fe().system_to_base_index(dof).second;
+            // get the unit support points for the relevant element
+            std::vector< Point<dim> > my_support_points = simulator_access.get_fe().base_element(base).get_unit_support_points();
+            if ( my_support_points.size() == 0 )
+              {
+                //manufacture a support point, arbitrarily at cell center
+                if ( dim==2 )
+                  unit_support_points.push_back( Point<dim> (0.5,0.5) );
+                if ( dim==3 )
+                  unit_support_points.push_back( Point<dim> (0.5,0.5,0.5) );
+              }
+            else
+              {
+                unit_support_points.push_back(my_support_points[base_index]);
+              }
+          }
+        return unit_support_points;
+      }
+  }
+
+  /**
+   * A set of helper functions that either return the point passed to it (if
+   * the current dimension is the same) or return a dummy value (otherwise).
+   */
+  namespace
+  {
+    const Point<2> as_2d(const Point<3> &/*p*/)
+    {
+      return Point<2>();
+    }
+
+    const Point<2> &as_2d(const Point<2> &p)
```

Comment:
Not wrong, but no need for the reference either
```suggestion
    Point<2> as_2d(const Point<2> &p)
```